### PR TITLE
Issue #261: General optimisation

### PR DESCRIPTION
**Pull request by tim-bezhashvyly**
Intermediate PR.

### CODE REVIEW

File: tests/Unit/Suites/Util/Factory/CommonFactoryTest.php [hunk 1]
```diff
@@ -862,65 +873,6 @@ public function testItCreatesAProductListingCanonicalTagSnippetKeyGenerator()
...
-            [ProductDetailViewSnippetRenderer::class],
-            [ProductInListingSnippetRenderer::class],
-            [PriceSnippetRenderer::class],
-            [ProductJsonSnippetRenderer::class],
-            [ConfigurableProductJsonSnippetRenderer::class],
-            [ProductCanonicalTagSnippetRenderer::class],
-            [ProductDetailPageRobotsMetaTagSnippetRenderer::class],
-        ];
-    }
-
```
Comment (Vinai) [inline tests/Unit/Suites/Util/Factory/CommonFactoryTest.php]: Do you think no longer testing which renderers are passed to the projector is a risk?

Comment (Vinai) [inline tests/Unit/Suites/Util/Factory/CommonFactoryTest.php]: As discussed on Slack, having a method return the renderers as an array improves testability and third party factory extendability.
However, passing them on as a variadic argument list is nice.
Can you partially revert the change accordingly?

File: src/Import/ContentBlock/ContentBlockProjector.php [hunk 2]
```diff
@@ -5,40 +5,48 @@
...
+     */
+    private function getSnippets(ContentBlockSource $projectionData): array
+    {
+        return array_reduce(
+            $this->snippetRenderers,
+            function ($carry, SnippetRenderer $snippetRenderer) use ($projectionData) {
+                return array_merge($carry, $snippetRenderer->render($projectionData));
+            },
+            []
+        );
```
Comment (tim-bezhashvyly) [inline src/Import/ContentBlock/ContentBlockProjector.php]: This part is duplicated on almost every projector. I'm thinking about injecting a Generic projector into non-generic ones instead. What do you think?

Comment (Schrank) [inline src/Import/ContentBlock/ContentBlockProjector.php]: Is it harder afterwards to change this in a project?

Comment (tim-bezhashvyly) [inline src/Import/ContentBlock/ContentBlockProjector.php]: Not sure I get your question.

Comment (Schrank) [inline src/Import/ContentBlock/ContentBlockProjector.php]: It sounds like more complexity to have less code and if every projector solves some other problem we get coupling on a place were code duplication might be better.

Comment (tim-bezhashvyly) [inline src/Import/ContentBlock/ContentBlockProjector.php]: This is what I'm asking. Do **you** this duplication is better here?

Comment (Schrank) [inline src/Import/ContentBlock/ContentBlockProjector.php]: It sounds to me two solved problems, so duplication is good here :)

File: tests/Unit/Suites/Import/GenericProjectorTest.php [hunk 3]
```diff
@@ -17,33 +18,35 @@ class GenericProjectorTest extends TestCase
...
     public function testSnippetIsWrittenIntoDataPool()
     {
-        $projectionSourceDataJson = '{}';
+        $testProjectionSourceData = 'foo';
+
+        $stubSnippetA = $this->createMock(Snippet::class);
+        $stubSnippetRendererA = $this->getMockBuilder(SnippetRenderer::class)
+            ->setMethods(['render'])
+            ->getMock();
+        $stubSnippetRendererA->method('render')->with($testProjectionSourceData)->willReturn([$stubSnippetA]);
```
Comment (Vinai) [inline tests/Unit/Suites/Import/GenericProjectorTest.php]: How about extracting a method `createStubSnippetRenderer()`?
It could be used by the previous test, too.

File: tests/Unit/Suites/Import/Product/ProductProjectorTest.php [hunk 4]
```diff
@@ -62,53 +48,65 @@ class ProductProjectorTest extends TestCase
...
+        $stubSnippetRendererA = $this->getMockBuilder(SnippetRenderer::class)
+            ->setMethods(['render'])
+            ->getMock();
+        $stubSnippetRendererA->method('render')->with($stubProductView)->willReturn([$stubSnippetA]);
+
+        $stubSnippetB = $this->createMock(Snippet::class);
+        $stubSnippetRendererB = $this->getMockBuilder(SnippetRenderer::class)
+            ->setMethods(['render'])
+            ->getMock();
+        $stubSnippetRendererB->method('render')->with($stubProductView)->willReturn([$stubSnippetB]);
```
Comment (Vinai) [inline tests/Unit/Suites/Import/Product/ProductProjectorTest.php]: As in the previous test, creating the stubSnippetRenderers adds a lot of noise to the setup section of the test. Maybe you can extract that into a utility method, too?

File: tests/Unit/Suites/ProductListing/Import/ProductListingSnippetProjectorTest.php [hunk 5]
```diff
@@ -18,75 +18,66 @@
...
+        $stubSnippetRendererA = $this->getMockBuilder(SnippetRenderer::class)
+            ->setMethods(['render'])
+            ->getMock();
+        $stubSnippetRendererA->method('render')->with($stubProductListing)->willReturn([$stubSnippetA]);
+
+        $stubSnippetB = $this->createMock(Snippet::class);
+        $stubSnippetRendererB = $this->getMockBuilder(SnippetRenderer::class)
+            ->setMethods(['render'])
+            ->getMock();
+        $stubSnippetRendererB->method('render')->with($stubProductListing)->willReturn([$stubSnippetB]);
```
Comment (Vinai) [inline tests/Unit/Suites/ProductListing/Import/ProductListingSnippetProjectorTest.php]: Same comment as above. Maybe it's worth creating a shared snippet stub renderer utility factory to be shared by the tests?

File: src/Import/ContentBlock/ContentBlockProjector.php (hunk 2 above)
Comment (Vinai) [inline src/Import/ContentBlock/ContentBlockProjector.php]: @Schrank I don't understand what you mean by "two solved problems" and "duplication is good here" or how in this case the duplication would make future changes simpler.

In my point of view in this case the reasons to remove duplication by injecting the `GenericProjector` outweigh the reasons to keep the duplication.

### Reasons for duplication:
* One less dependency per projector.
* Less work (e.g. keep that part of the PR in the current shape).
* All the logic for rendering and combining the results of the snippet renderer array is contained within each projector.

### Reasons to remove duplication
* In this case the duplicate code represents is the same concept, not two concepts from different contexts that happen to be implmented by the same code.
  The "idea" that is implemented multiple times here is collecting the results of snippet renderers into a single array.
* Because of the above, removing duplication will make the code easier to change in future. If the snippet renderer API changes, we currently would have to adjust all projectors. When the duplication is removed, only one class will be affected (the `GenericProjector`).
* The delegation to the `GenericProjector` is an implementation detail of the other projectors, that is, it doesn't leak to the clients using the respective projectors.  
  That means if in future a specific projector needs to work differently, the delegation to the `GenericProjector` can be removed and replaced with specific logic without affecting any other projectors.  
  This is the beauty of composition over inheritance and encapsulation.

To summarize, I support your idea to remove the duplication.
